libs/libc/machine/arch_atomic.c

```diff
@@ -733,3 +733,26 @@ SYNC_VAL_CMP_SWAP(4, uint32_t)
 SYNC_VAL_CMP_SWAP(8, uint64_t)
 
 #endif /* __clang__ */
+
+/****************************************************************************
+ * Name: up_testset
+ ****************************************************************************/
+
+#if defined(CONFIG_SPINLOCK) && !defined(CONFIG_ARCH_HAVE_TESTSET)
+spinlock_t up_testset(volatile FAR spinlock_t *lock)
+{
+  irqstate_t flags;
+  spinlock_t ret;
+
+  flags = spin_lock_irqsave(NULL);
```

Comment:
@xiaoxiang781216 are you sure in this change? The first look makes me thinking that change from `up_irq_save` to `spin_lock_irqsave` will make a recursion.

Reply:
I mean `up_testset` -> `spin_lock_irqsave` -> `spin_lock` -> `up_testset`

Reply:
This call chain only happen in SMP case, but SMP require arch provide up_testset implement(ARCH_HAVE_TESTSET):
https://github.com/apache/incubator-nuttx/blob/master/sched/Kconfig#L275
so, the implementation in here is only used for single core.

Reply:
So why we can't keep up_irq_save?

Reply:
just align with the code at:
https://github.com/apache/incubator-nuttx/pull/7649/files#diff-60c271175313a6344e1faeba2708393238fd3caf3cf282b3f247d9a88699cba9R67

Reply:
Sorry, that link does not redirect me to anything meaningful

Reply:
Here:
<img width="359" alt="image" src="https://user-images.githubusercontent.com/18066964/203399052-02811b6a-492b-4256-8408-10da8a206337.png">

Reply:
Yeah, those are atomic values, but spinlock itself relies at up_testest, so using spinlock inside seems to be odd at least to me.
I fully understand that spinlock can be implemented on top of atomic compare-and-swap, but this is not the case here. Also the macro for a single CPU system will end up into disabling of interrupts as it was in the code before this change.
@masayuki2009 what is your opinion here?

Reply:
>@xiaoxiang781216 are you sure in this change? The first look makes me thinking that change from up_irq_save to spin_lock_irqsave will make a recursion.

I think so too.

Reply:
Ok, let's keep it as inline function.